### PR TITLE
New API Metrics, and start of a libtsapi.so structure

### PR DESCRIPTION
**Pull request by zwoop**
The new API metrics are backwards compatible with previous APIs, since the old API wraps these new APIs. New plugins / code could chose to use the Metrics.h API directly.

Work to be finished here includes

   1. Decide what goes into include/api ? Do we have one new include file to #include all the include/api/*.h files?

   2. What should be moved to src/api and include/api.

Lost in this merge is all the work from Chris McFarlen. Thanks!

Co-Author: Chris McFarlen <chris@mcfarlen.us>

### CODE REVIEW

**Comment by zwoop** [comment]
[approve ci autest]

**Comment by zwoop** [comment]
I did some quick tests using the remap_stats.so plugin (a global plugin), there's no noticeable difference in performance even though I'm just putting pressure on 3 single metrics (each updated on every request). 

New Metrics:
```
100 mean bytes/fetch
1167275.00 fetches/sec, 1.16728e+08 bytes/sec
msecs/connect: 0.084 mean, 0.366 max, 0.027 min
msecs/first-response: 0.476 mean, 50.357 max, 0.024 min
```

Old Metrics:

```
100 mean bytes/fetch
1145004.00 fetches/sec, 1.14500e+08 bytes/sec
msecs/connect: 0.083 mean, 0.373 max, 0.025 min
msecs/first-response: 0.482 mean, 46.571 max, 0.023 min
```

File: src/traffic_logcat/Makefile.inc [hunk 1]
```diff
@@ -47,7 +47,8 @@ traffic_logcat_traffic_logcat_LDADD = \
 	$(top_builddir)/iocore/utils/libinkutils.a \
 	$(top_builddir)/src/tscore/libtscore.la \
 	$(top_builddir)/src/tscpp/util/libtscpputil.la \
-        $(top_builddir)/mgmt/utils/libutils_p.la
+    $(top_builddir)/mgmt/utils/libutils_p.la \
+    $(top_builddir)/src/api/libtsapi.la
```
Comment (bneradt) [inline src/traffic_logcat/Makefile.inc:51]: nit: looks like inconsistent spacing. The previous lines are tabs, these new lines are spaced.

File: src/traffic_logstats/Makefile.inc [hunk 2]
```diff
@@ -52,7 +52,8 @@ traffic_logstats_traffic_logstats_LDADD = \
 	$(top_builddir)/iocore/utils/libinkutils.a \
 	$(top_builddir)/src/tscore/libtscore.la \
 	$(top_builddir)/src/tscpp/util/libtscpputil.la \
-        $(top_builddir)/mgmt/utils/libutils_p.la
+    $(top_builddir)/mgmt/utils/libutils_p.la \
+    $(top_builddir)/src/api/libtsapi.la
```
Comment (bneradt) [inline src/traffic_logstats/Makefile.inc:56]: Spacing.

**Review by bneradt — CHANGES_REQUESTED** [review]

**Comment by ezelkow1** [comment]
[approve ci]

**Comment by zwoop** [comment]
[approve ci rat]

**Comment by zwoop** [comment]
[approve ci]

File: src/traffic_logcat/Makefile.inc (hunk 1 above)
Comment (zwoop) [reply src/traffic_logcat/Makefile.inc]: I fixed these.

**Comment by zwoop** [comment]
@bneradt can you review again please ?

**Review by bneradt — DISMISSED** [review]

**Comment by bryancall** [comment]
[approve ci clang-analyzer]

**Review by bneradt — DISMISSED** [review]

**Comment by bneradt** [comment]
[approve ci autest]

**Review by bneradt — APPROVED** [review]

**Comment by zwoop** [comment]
[approve ci autest]

**Comment by bneradt** [comment]
[approve ci autest]

**Pull request merged**